lib/irc/Scheduler.js

```diff
@@ -0,0 +1,107 @@
+/*eslint no-invalid-this: 0*/
+"use strict"
+var promiseutil = require("../promiseutil");
+var Promise = require("bluebird");
+var logging = require("../logging");
+var log = logging.get("scheduler");
+
+function ScheduledQueue (server) {
+    this._queue = [];
+    this._processing = null;
+    this._server = server;
+
+    setInterval(this._consume.bind(this), server.getReconnectInterval());
+}
+
+ScheduledQueue.prototype._consume = Promise.coroutine(function*() {
+    if (this._processing) {
+        return;
+    }
+    this._processing = this._queue.shift();
+    if (!this._processing) {
+        return;
+    }
+    try {
+        let thing = this._procFn(this._processing);
+
+        let result = yield thing;
+        log.info(`Resolving scheduled promise for ${this._server.domain}`);
+        this._processing.defer.resolve(result);
+    }
+    catch (err) {
+        log.info(`Rejecting scheduled promise for ${this._server.domain} (${err.message})`);
+        this._processing.defer.reject(err);
+    }
+    finally {
+        this._processing = null;
+    }
+});
+
+ScheduledQueue.prototype._procFn = function (item) {
+    return item.fn();
+}
+
+ScheduledQueue.prototype.enqueue = function(item) {
+    this._queue.push(item);
+}
+
+ScheduledQueue.prototype.killAll = function() {
+    for (var i = 0; i < this._queue.length; i++) {
+        this._queue[i].reject();
+    }
+}
+
+/**
+ * An IRC connection scheduler. Enables ConnectionInstance to reconnect
+ * in a way that queues reconnection requests and services the FIFO queue at a
+ * rate determined by ircServer.getReconnectInterval().
+ */
+
+var Scheduler = {
+    _queueServers: [],
+    _queues: {},
+    _getQueue: _getQueue,
```

Comment:
I don't like the amount of indirection here. Why not just in-line this, or call the global function directly?